File: go/blind75/arraysAndHashing/topKFrequentElements.go
package arraysAndHashing

// 347. Top K Frequent Elements (Medium)
func TopKFrequent(nums []int, k int) (res []int) {
	countMap := make(map[int]int)
	countSlice := make([][]int, len(nums)+1)

	for _, num := range nums {
		if count, ok := countMap[num]; !ok {
			countMap[num] = 1
		} else {
			countMap[num] = count + 1
		}
	}

	for num, count := range countMap {
		countSlice[count] = append(countSlice[count], num)
	}

	for i := len(countSlice) - 1; i > 0; i-- {
		res = append(res, countSlice[i]...)
		if len(res) == k {
			return
		}
	}

	return
}

File: python/blind_75/arrays_and_hashing/top_k_frequent_elements.py
from typing import List
import unittest

# 347. Top K Frequent Elements (Medium)
class Solution:
  def top_k_frequent(self, nums: List[int], k: int) -> List[int]:
    count = {}
    freq = [[] for i in range(len(nums) + 1)]

    for n in nums:
      count[n] = 1 + count.get(n, 0)
    for n, c in count.items():
      freq[c].append(n)

    res = []
    for i in range(len(freq) - 1, 0, -1):
      for n in freq[i]:
        res.append(n)
        if len(res) == k:
          return res

class Test(unittest.TestCase):
  def test_1(self):
    self.assertEqual(Solution.top_k_frequent(self, [1,1,1,2,2,3], 2), [1,2], "Should be [1,2]")

  def test_2(self):
    self.assertEqual(Solution.top_k_frequent(self, [1], 1), [1], "Should be [1]")

if __name__ == '__main__':
  unittest.main()
